comm/src/main/scala/io/casperlabs/comm/gossiping/SynchronizerImpl.scala

```
@@ -0,0 +1,247 @@
+import cats.implicits._
+import com.google.protobuf.ByteString
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.auto._
+import eu.timepit.refined.numeric._
+import io.casperlabs.casper.consensus.BlockSummary
+import io.casperlabs.comm.discovery.Node
+import io.casperlabs.crypto.codec.Base16
+import io.casperlabs.shared.Log
+
+import scala.collection.immutable.Queue
+import scala.util.control.NoStackTrace
+
+// TODO: Optimise to heap-safe
+class SynchronizerImpl[F[_]: Sync: Log](
+    connectToGossip: Node => F[GossipService[F]],
+    backend: Synchronizer.Backend[F],
+    maxPossibleDepth: Int Refined Positive,
+    maxPossibleWidth: Int Refined Positive,
+    maxDepthAncestorsRequest: Int Refined Positive
+) extends Synchronizer[F] {
+  import io.casperlabs.comm.gossiping.SynchronizerImpl._
+
+  override def syncDag(
+      source: Node,
+      targetBlockHashes: Set[ByteString]
+  ): F[Vector[BlockSummary]] = {
+    val sync = for {
+      service        <- connectToGossip(source)
+      tips           <- backend.tips
+      justifications <- backend.justifications
+      syncState      <- loop(service, targetBlockHashes.toList, tips ::: justifications, SyncState.empty)
+      missing        <- missingDependencies(syncState.dag)
+      newDagInTopologicalOrder <- if (missing.isEmpty)
+                                   topologicalSort(syncState, targetBlockHashes).pure[F]
+                                 else
+                                   Log[F]
+                                     .warn(s"Streamed DAG contains missing dependencies: ${missing
+                                       .map(h => Base16.encode(h.toByteArray).take(6))}")
+                                     .map(_ => Vector.empty[BlockSummary])
+    } yield newDagInTopologicalOrder
+
+    sync.handleErrorWith { error =>
+      Log[F]
+        .error(s"Failed to stream ancestor block summaries, ${error.getMessage}", error)
+        .map(_ => Vector.empty[BlockSummary])
```

It would be better to return the error and let the caller decide what to do, rather then return an empty DAG. For example this way the `NewBlocks` handler will end up logging "Synced 0 blocks with sender.". We might want to do different action based on why the syncing failed, whether it was a timeout or problems with the data.

---

Done, now [it returns explicit errors](https://github.com/CasperLabs/CasperLabs/pull/379/commits/e63235d8495ae3c75baea95832cadda66c7d706a)